fili-core/src/main/java/com/yahoo/bard/webservice/table/availability/BaseCompositeAvailability.java

```diff
@@ -0,0 +1,60 @@
+// Copyright 2017 Yahoo Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.table.availability;
+
+import com.yahoo.bard.webservice.data.config.names.TableName;
+import com.yahoo.bard.webservice.util.SimplifiedIntervalList;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * A class implementing common capabilities for availabilities backed by a collection of other availabilities.
+ */
+public abstract class BaseCompositeAvailability implements Availability {
+
+    /**
+     * Return a stream of all the dependent availabilities.
+     *
+     * @return A stream of availabilities
+     */
+    protected abstract Stream<Availability> getAllDependentAvailabilities();
+
+    @Override
+    public Set<TableName> getDataSourceNames() {
+        return getAllDependentAvailabilities()
```

Comment:
If this class takes the collection of dependent availabilities, this calculation can be done in the constructor once, rather than each time this method is called.